strutil/quantity/quantity.go

```diff
@@ -200,3 +201,248 @@ func FormatDuration(dt float64) string {
+)
+
+// Space delimit options
+type SpaceMode bool
+
+const (
+	SpaceOff SpaceMode = false
+	SpaceOn  SpaceMode = true
+)
+
+// The minimum Duration parameter allows the rendering to discard the units
+// of time less than the minimum. However, depending on what we are
+// rendering, we either need to peform a ceiling, floor or rounding
+// operation with the remainder below the minimum.
+type RenderMode uint32
+
+const (
+	TimeLeft RenderMode = iota
+	TimePassed
+	TimeRounded
+)
+
+// FormatDurationGeneric formats time duration similar to systemd
+// format_timespan, but with options allowing for a more compact
+// arrangement, including shortened suffixes. The function is provided
+// with dt in seconds (as in the output of time.Now().Seconds()). The 'min'
+// and 'max' Duration allows limiting the unit range rendered. Place values
+// less than 'min' will be processed according to RenderMode. RenderMode
+// allows rounding, flooring or ceiling to be applied to place value below
+// 'min'. Durations above the 'max' place value will result in "ages!".
+func FormatDurationGeneric(dt float64, min Duration, max Duration, count Places, mode RenderMode, space SpaceMode) string {
+	var units = map[Duration]string{
+		Year:    "y",
+		Day:     "d",
+		Hour:    "h",
+		Minute:  "m",
+		Second:  "s",
+		MSecond: "ms",
+		USecond: "µs",
+		NSecond: "ns",
+	}
+
+	// We need to access the map in order using an ordered array
+	var keys []Duration
+	for k := range units {
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] > keys[j] })
+
+	var render string
+
+	// Invalid case: min > max
+	if min > max {
+		return "inv!"
+	}
+
+	// Special case: zero duration
+	if dt <= 0 {
+		return "0" + units[min]
+	}
+
+	// Special case: render as "ages!"
+	if dt >= float64(math.MaxUint64) {
+		return "ages!"
+	}
+
+	// Fractional seconds to nanoseconds
+	delta := Duration(dt * float64(Second))
+
+	// If we only render a subset of place values, we need to apply
+	// the render mode to the last place value. We do this indirectly
+	// by moving up the 'min' accuracy.
+	if count == ShowCompact {
+		for i, key := range keys {
+			// Compact mode only renders two places up until seconds
+			// The indexing is safe as places beyond seconds exist
+			if delta >= key && key >= Minute {
+				index := i + 1
+				if keys[index] > min {
+					min = keys[index]
+				}
+				break
+			}
+
+			// Only one place if we are s, ms, us or ns
+			if delta >= key && key < Minute {
+				if key > min {
+					min = key
+				}
+				break
+			}
+		}
+	}
+
+	// Apply the rendermode to the least significant place value
+	mod := delta % min
+	switch {
+	case mode == TimeLeft && mod > 0:
+		// Ceiling
+		delta = delta + min - mod
+	case mode == TimePassed:
+		// Floor
+		delta = delta - mod
+	case mode == TimeRounded && mod >= (min/2):
+		// Rounded Up
+		delta = delta + min - mod
+	case mode == TimeRounded && mod < (min/2):
+		// Rounded Down
+		delta = delta - mod
```

Comment:
nitpick: using `+=` and `-=` IMHO is more readable

Reply:
Fixed.